planner/core/memtable_predicate_extractor.go

```diff
@@ -0,0 +1,204 @@
+	SkipRequest bool
+
+	// NodeTypes represents all components types we should send request to.
+	// e.g:
+	// 1. SELECT * FROM tidb_cluster_config WHERE type='tikv'
+	// 2. SELECT * FROM tidb_cluster_config WHERE type in ('tikv', 'tidb')
+	NodeTypes set.StringSet
+
+	// Addresses represents all components addresses we should send request to.
+	// e.g:
+	// 1. SELECT * FROM tidb_cluster_config WHERE address='192.168.1.7:2379'
+	// 2. SELECT * FROM tidb_cluster_config WHERE type in ('192.168.1.7:2379', '192.168.1.9:2379')
+	Addresses set.StringSet
+}
+
+// Extract implements the MemTablePredicateExtractor Extract interface
+func (e *ClusterConfigTableExtractor) Extract(schema *expression.Schema, names []*types.FieldName, predicates []expression.Expression) []expression.Expression {
+	remained := make([]expression.Expression, 0, len(predicates))
+	// All columns can be pushed down to the memory table `TIDB_CLUSTER_CONFIG`
+	const (
+		ColNameType    = "type"
+		ColNameAddress = "address"
+	)
+	extractCols := make(map[int64]*types.FieldName)
+	for i, name := range names {
+		if ln := name.ColName.L; ln == ColNameType || ln == ColNameAddress {
+			extractCols[schema.Columns[i].UniqueID] = name
+		}
+	}
+	// We don't support change the schema of memory table, so something must
+	// be wrong if the `type/address` columns cannot be found in the schema.
+	// The purpose of the following assert is used to make test failure because
+	// it will never happen if our release passing all test cases.
+	if len(extractCols) != 2 {
+		panic(fmt.Sprintf("push down columns `type/address` not found in schema, got: %+v", extractCols))
+	}
+
+	e.NodeTypes = set.NewStringSet()
+	e.Addresses = set.NewStringSet()
+
+	// We should use INTERSECTION of sets because of the predicates is CNF array
+	for _, expr := range predicates {
+		var extracted bool
+		switch x := expr.(type) {
+		case *expression.ScalarFunction:
+			switch x.FuncName.L {
+			case ast.EQ:
+				colName, datum := e.extractColEqConsExpr(extractCols, x)
+				switch colName {
+				case ColNameType:
+					tp := datum.GetString()
+					// SELECT * FROM tidb_cluster_config WHERE type='a' AND type='b'
+					if len(e.NodeTypes) > 0 && !e.NodeTypes.Exist(tp) {
+						e.NodeTypes = set.NewStringSet()
+						e.SkipRequest = true
+						return nil
+					}
+					e.NodeTypes.Insert(tp)
+				case ColNameAddress:
+					addr := datum.GetString()
+					// SELECT * FROM tidb_cluster_config WHERE address='a' AND address='b'
+					if len(e.Addresses) > 0 && !e.Addresses.Exist(addr) {
+						e.Addresses = set.NewStringSet()
+						e.SkipRequest = true
+						return nil
+					}
+					e.Addresses.Insert(addr)
+				}
+				extracted = colName != ""
+			case ast.In:
+				colName, datums := e.extractColInConsExpr(extractCols, x)
+				switch colName {
+				case ColNameType:
+					tmpNodeTypes := set.NewStringSet()
+					for _, datum := range datums {
+						tmpNodeTypes.Insert(datum.GetString())
```

Comment:
Can we assert the datum type is always string?

Reply:
All datum can be converted to a string (node type and address always be a string here).